File: maxhanna.client/src/services/datacontracts/social/poll.ts
export class Poll {
  id!: number;
  componentId!: string;  // Matches the story ID or other component identifier
  question!: string;
  options!: PollOption[];
  userVotes!: PollVote[];  // Votes from all users
  currentUserVote?: string; // The current user's vote (if any)
  totalVotes!: number;
  expiresAt?: Date;
  createdAt!: Date;
  isClosed!: boolean;

  constructor(data?: Partial<Poll>) {
    if (data) {
      Object.assign(this, data);

      // Ensure proper Date objects
      if (data.expiresAt) this.expiresAt = new Date(data.expiresAt);
      if (data.createdAt) this.createdAt = new Date(data.createdAt);

      // Initialize arrays if not provided
      this.options = data.options || [];
      this.userVotes = data.userVotes || [];

      // Calculate total votes
      this.totalVotes = this.userVotes.length;
    }
  }
}

export interface PollOption {
  id: string;  // Unique identifier for the option
  text: string;
  voteCount: number;
  percentage?: number; // Calculated percentage of votes
}

export interface PollVote {
  id: number;
  userId: number;
  componentId: string;
  value: string;  // Matches the option id
  timestamp: Date;
  username?: string; // Optional user info
  displayPicture?: string; // Optional user avatar
}

File: maxhanna.Server/Controllers/DataContracts/Social/Poll.cs
namespace maxhanna.Server.Controllers.DataContracts.Social
{
	public class Poll
	{
		public string ComponentId { get; set; } = string.Empty;
		public string Question { get; set; } = string.Empty;
		public List<PollOption> Options { get; set; } = new List<PollOption>();
		public List<PollVote> UserVotes { get; set; } = new List<PollVote>();
		public int TotalVotes { get; set; }
		public DateTime CreatedAt { get; set; }
	}

	public class PollOption
	{
		public string Id { get; set; } = string.Empty; // Unique identifier for the option (e.g., "1", "2", or option text)
		public string Text { get; set; } = string.Empty; // Display text for the option
		public int VoteCount { get; set; }
		public int Percentage { get; set; }
	}

	public class PollVote
	{
		public int Id { get; set; }
		public int UserId { get; set; }
		public string ComponentId { get; set; } = string.Empty;
		public string Value { get; set; } = string.Empty; // The selected option's ID or text
		public DateTime Timestamp { get; set; }
		public string Username { get; set; } = string.Empty;
		public string? DisplayPicture { get; set; }
	}
}